<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Klaudia Chmielewska" />


<title>ML in chem</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>







<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Machine Learning in Chemistry</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Main</a>
</li>
<li>
  <a href="session1.html">Lesson 1</a>
</li>
<li>
  <a href="session2.html">Lesson 2</a>
</li>
<li>
  <a href="session3.html">Lesson 3</a>
</li>
<li>
  <a href="session4.html">Lesson 4</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore"><strong>ML in chem</strong></h1>
<h3 class="subtitle"><em>Lesson 4</em></h3>
<h4 class="author"><strong><em>Klaudia Chmielewska</em></strong></h4>

</div>


<div id="partial-least-squares" class="section level1">
<h1>Partial Least Squares</h1>
<p><br />
</p>
<p>On our last class we studied Multiple Linear Regression. Linear
Regression itself, allows us to draw a line of dependency between 1
variable and the other 1 (between 2 columns of data, between 2
characteristics). Multiple one considered that more than 1 variable
affects one another.</p>
<p>But what if you have a lot of <em>independent</em> variables - many
of which are correlated to other independent variables…? This means we’d
end up with so called multicollinearity, where several independent
variables are correlated (with correlation coefficient = 1 when relation
is perfectly colinear). It’s crucial to understand that such
multicopllinearity can happen when we have a huge number of predictors,
but low amount of observations (like in the case of nanomaterials, which
is close to some of you!).</p>
<p>This however undermines the theory of regression, stating about the
dependency between variables. The solution in such situation can be
Partial Least Squares, which reduces the dimensionality of correlated
variables and model underlying correlations.</p>
<p>If you got a feeling that it seems like an unsupervised group ypou
are partially correct - as PLS combines features from PCA (principal
component analysis) and multiple regression! Hence taking something from
both worlds!</p>
<p>Last Lesson, you were presented with an example on wine quality.</p>
<p>For such, the predictors can be price and sugar/alcohol/acidity
content.</p>
<p>If we checked the correlation between said predictors we might see
whether the price correlated negatively with the quality or not, or
alcohol content, etc.</p>
<p>Knowing there is no basic linear regression between price and
quality, PLS offers a more precise analysis, taking into consideration
less-obvious aspects.</p>
<div id="applicability" class="section level2">
<h2>Applicability</h2>
<p><br />
</p>
<p>PLS is useful in drug / food / plastic industries, as well as
scientific communities of course. PLS is however not usually used in
screening out variables that are not useful in explaining the response,
which means it’s not good for finding “secret relationships”.</p>
<p>Nevertheless it will be good to develop predictive models, analyzing
how spectral measurements (NIR, IR, UV) correlate to chemical
composition or other physio-chemical properties, which could lead to a
better computational approximation of the spectra for the yet
unsynthesized substances, or to predict the spectra in order to help the
scientist assess whether he got an appropriate result (like in
MestReNova, where you can ask the program to propose a spectra based on
the chemical composition).</p>
<p>PLS regression can preserve the asymmetry (differences/undependency)
of the variables and predictors! Whereas many other techniques do not!
This is a huge value of PLS!!</p>
<p><strong>However, one must remember using PLS - being multivariate,
the results can significantly differ from those calculated for each of
the variables separately!!!</strong></p>
<p>I believe the best laurel to PLS in chemistry is <a
href="https://doi.org/10.1007/s12045-021-1140-1">this</a>. Feel free to
read it to gain broader perspective on the method!</p>
<p>Additionally, <a
href="https://github.com/selbouhaddani/OmicsPLS_ShortCourse">here</a>
you can find a short course, using R.</p>
<p>Both R and Python have dedicated packages/libraries for PLS for
different application. Here are ones for you to consider: - <a
href="https://github.com/selbouhaddani/OmicsPLS">R omics</a> - <a
href="https://github.com/cran/asmbPLS">R patient phenotypes using
multi-omics</a> - <a
href="https://github.com/tanghaibao/goatools">Python gene
onthology</a></p>
<p>Example of a ML project using different algorithms, including PLS: <a
href="https://github.com/EmanOssama/Predict-Song-Popularity-Machine-Learning-Project">Song
prediction</a></p>
</div>
<div id="exercise-1" class="section level2">
<h2>Exercise 1</h2>
<p>Pick 1 publication and describe how they used PLS in its topic of
interest.</p>
<p>Spectral data: <a
href="https://doi.org/10.1007/s12393-016-9147-1">1</a> <a
href="https://doi.org/10.1038/s41598-022-27112-y">2</a> <a
href="https://doi.org/10.3389/fcell.2020.00385">3</a> <a
href="https://doi.org/10.1021/acs.analchem.0c00075">4</a></p>
<p>Omics: <a href="https://doi.org/10.1016/j.aca.2015.02.012">1
metabolomics</a> <a href="https://doi.org/10.1089/omi.2021.0201">2 blood
pressure</a> <a href="https://doi.org/10.1111/jth.14875">3
transcriptomics</a> <a
href="https://doi.org/10.1038/s41598-023-33194-z">4 proteomics</a> <a
href="https://doi.org/10.1186/s12967-022-03311-z">5 metabolomics
cancer</a> <a href="https://doi.org/10.1016/j.phymed.2022.154635">6
proteomics lung cancer</a> <a
href="https://doi.org/10.1155/2021/3291762">7 lymphoma</a> <a
href="doi.org/10.1093/bioinformatics/btv535">8 genomics</a> <a
href="10.3389/fphar.2023.1200795">9 glioma genomics</a> <a
href="https://doi.org/10.3389/fphys.2022.953237">10 obesity</a></p>
<p>QSAR: <a href="https://doi.org/10.1016/j.molstruc.2020.129156">1
neuropathic pain</a> <a
href="https://doi.org/10.1007/s00894-021-04906-x">2 antidepressants</a>
<a href="https://doi.org/10.1007/s11030-022-10596-1">3 virtual
screening</a> <a href="https://doi.org/10.1016/j.molstruc.2021.130571">4
corrosion</a> <a
href="https://doi.org/10.2174/1570180819666220420092723">5
mycobacterium</a> <a
href="https://doi.org/10.1016/j.scitotenv.2023.166708">6 aquatic
toxicity</a> <a href="https://doi.org/10.1007/s10953-021-01083-6">7 tau
- very interesting for me!</a> <a
href="https://doi.org/10.1016/j.memsci.2022.120268">8 nanofiltration</a>
<a href="https://doi.org/10.1016/j.chemosphere.2021.130652">9 ecotox</a>
<a href="10.1039/D1EN00733E">10 nanoparticles</a> <a
href="https://doi.org/10.1021/acsomega.0c01323">11 2D QSAR &amp;
docking</a> <a href="https://doi.org/10.1007/s11051-019-4541-2">12
nanoparticle toxicity ranking in small datasets - very good too!</a></p>
</div>
<div id="exercise-2" class="section level2">
<h2>Exercise 2</h2>
<p>Pick 1 of the following repositories, read their codes, copy its
parts to create your own code. You can work in groups. In the end, based
on the publication, repositories and your own knowledge, you should have
your own exemplatory code for PLS on your dataset of interest.</p>
<p><a href="https://github.com/tkimhofer/opls">MS, NMR</a> <a
href="https://github.com/victorpozzobon/pigmentQuantificationSpectrophotometry">pigment
spectrophotometric quantification</a> <a
href="https://github.com/victorpozzobon/PSOGAPLS_NitrogenSpectroscopy">nitrite
spectroscopic quantification</a> <a
href="https://github.com/OzyKartike/BioInformatics">QSAR</a></p>
</div>
<div id="goatools" class="section level2">
<h2>goatools</h2>
<p><br />
</p>
<p>Let’s now focus on goatools in Python. Please create a Jupyter
notebook to work in it.</p>
<p>First, go through <a
href="https://github.com/dvklopfenstein/goatools_simulation">this
repository</a> in order to understand the possibilities of goatools.</p>
<p>Now, open <a href="https://github.com/stsour/BIOE7374_Omics">the next
one</a>. You should see… Gene ontology once more! Now instead of using
david bioserver, you can also run scripts on you genes of interest
without the need if an online server.</p>
<p><em>Why is that important?</em> Because if you want to patent a given
substance, it should not be submitted to online servers before
submission to the Patent Office!!!!!</p>
<p>Go through the proteomic scripts in that last link carefully. Look
through both the code and the easthetics of the csv file.</p>
</div>
<div id="exercise-3" class="section level2">
<h2>Exercise 3</h2>
<p><br />
</p>
<p>Use <a href="https://github.com/stsour/BIOE7374_Omics">that
repository</a>. Edit its components (scripts &amp; csv files) and run
your own calculation and PLS. For data you can either use a cut version
of the dataset presented in the repository or choose one of your own
(maybe your old dataset from Omics class?). Prepare a Jupyter notebook
with the code, comments, graphs etc. Eventually you can send me the code
and results description separately.</p>
<p>Send finished exercises (Word file+code file or Jupyter notebook or
Rmarkdown) to my email: <strong><a
href="mailto:klaudia.chmielewska@ug.edu.pl"
class="email">klaudia.chmielewska@ug.edu.pl</a></strong> with the Title
format: “DCH2_Name_Class4” (DCH = Digital Chemistry)</p>
<hr />
<p><br />
<br />
Klaudia Chmielewska</p>
<p><em><a href="mailto:klaudia.chmielewska@ug.edu.pl"
class="email">klaudia.chmielewska@ug.edu.pl</a></em></p>
<p><br />
<br />
</p>
<p>See you in the next lesson!</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
